raiden_contracts/contracts/services/UserDeposit.sol

```diff
@@ -0,0 +1,142 @@
+    }
+
+    /*
+     *  Events
+     */
+
+    event BalanceReduced(address indexed owner, uint newBalance);
+    event WithdrawPlanned(address indexed withdrawer, uint plannedBalance);
+
+    /*
+     *  Modifiers
+     */
+
+    modifier canTransfer() {
+        // TODO: allow only access from MSC and 1-n payment contract
+        require(true);
+        _;
+    }
+
+    /*
+     *  Constructor
+     */
+
+    /// @notice Set the default values for the smart contract
+    /// @param _token_address The address of the token to use for rewards
+    constructor(
+        address _token_address
+    )
+        public
+    {
+        require(_token_address != address(0x0));
+        require(contractExists(_token_address));
+
+        token = Token(_token_address);
+        // Check if the contract is indeed a token contract
+        require(token.totalSupply() > 0);
+    }
+
+    /// @notice Deposit tokens. Idempotent function that sets the
+    /// total_deposit of tokens of the beneficiary.
+    /// Can be called by anyone several times and on behalf of other accounts
+    /// @param beneficiary The account benefiting from the deposit
+    /// @param total_deposit The sum of tokens, that have been deposited
+    function deposit(address beneficiary, uint256 total_deposit) public
+    {
+        require(total_deposit > balances[beneficiary]);
+
+        // Calculate the actual amount of tokens that will be transferred
+        uint256 added_deposit = total_deposit - balances[beneficiary];
+
+        balances[beneficiary] += added_deposit;
+        require(token.transferFrom(msg.sender, address(this), added_deposit));
+    }
+
+    /// @notice Internally transfer deposits between two addresses
+    /// The amount will be deducted from the msg sender's balance
+    /// @param receiver Account to which the amount will be credited
+    /// @param amount Amount of tokens to be withdrawn
+    function transfer(
+        address sender,
+        address receiver,
+        uint256 amount
+    )
+        canTransfer()
+        public returns (bool success)
+    {
+        if (balances[sender] >= amount && amount > 0) {
+            balances[sender] -= amount;
+            balances[receiver] += amount;
+            emit BalanceReduced(sender, balances[sender]);
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    /// @notice Announce intention to withdraw tokens.
+    /// Sets the planned withdraw amount and resets the earliest_withdraw
+    /// @param amount Maximum amount of tokens to be withdrawn
+    function planWithdraw(uint256 amount) public {
+        require(amount > 0);
+        require(balances[msg.sender] >= amount);
+
+        withdraw_plans[msg.sender] = WithdrawPlan({
+            amount: amount,
+            earliest_withdraw: block.number + withdraw_delay
+        });
+        emit WithdrawPlanned(msg.sender, balances[msg.sender] - amount);
```

Comment:
Just a note for `planWithdraw` - this function enables users to overwrite their previous planned withdraws and emits an event each time this is overwritten. I assume this is the intention.